### PR TITLE
Make loader app load updated system image directly

### PR DESCRIPTION
**Pull request by yuying-y**
b/344069914
b/328421395

### CODE REVIEW

**Comment by yuying-y** [comment]
Tested and verified on Linux

File: starboard/loader_app/slot_management.cc [hunk 1]
```diff
@@ -54,6 +62,103 @@ const char kCompressedCobaltLibraryName[] = "libcobalt.lz4";
...
+
+  struct stat info;
+  if (stat(manifest_file_path.data(), &info) != 0) {
+    SB_LOG(WARNING)
+        << "Failed to open the manifest file at the installation path.";
+    return false;
+  }
+
+  bool version_found = false;
+  std::ifstream manifest_file(std::string(manifest_file_path.data()));
```
Comment (y4vor) [inline starboard/loader_app/slot_management.cc:139]: Please use `third_party/jsoncpp` to parse the `manifest.json`.
Also adde unit tests to `starboard/loader_app/slot_management_test.cc`

File: starboard/loader_app/slot_management.cc [hunk 2]
```diff
@@ -14,8 +14,14 @@
 
 #include "starboard/loader_app/slot_management.h"
 
+#include <stdio.h>
 #include <sys/stat.h>
 
+#include <algorithm>
+#include <fstream>
```
Comment (kaidokert) [inline starboard/loader_app/slot_management.cc:21]: See my response in the chat, but avoid using `<fstream>`, this will lead to leaks. In here, you can probably use `starboard/common/file` to read in file content, and then pass it on to the JSON parser with a string in-memory.

File: starboard/loader_app/slot_management.cc (hunk 1 above)
Comment (yuying-y) [reply starboard/loader_app/slot_management.cc]: Done. Please review again.

File: starboard/loader_app/slot_management.cc (hunk 2 above)
Comment (yuying-y) [reply starboard/loader_app/slot_management.cc]: Thanks, yep, I used ScopedFile from `starboard/common/file` now.

File: starboard/loader_app/slot_management_test.cc [hunk 3]
```diff
@@ -468,6 +489,81 @@ TEST_P(SlotManagementTest, BadSabi) {
   SbFileDeleteRecursive(good_path.c_str(), false);
 }
 
+TEST_P(SlotManagementTest, CompareEvergreenVersion) {
+  if (!storage_path_implemented_) {
+    return;
+  }
+  std::vector<char> v1(kTestEvergreenVersion1,
+                       kTestEvergreenVersion1 + strlen(kTestEvergreenVersion1));
+  std::vector<char> v2(kTestEvergreenVersion2,
```
Comment (hlwarriner) [inline starboard/loader_app/slot_management_test.cc:498]: Since the details about `kTestEvergreenVersion2` are highly relevant to the test, I think it would be more readable if it were a local variable instead of a constant. Please see go/unit-testing-practices#constants.

File: starboard/loader_app/slot_management_test.cc [hunk 4]
```diff
@@ -468,6 +489,81 @@ TEST_P(SlotManagementTest, BadSabi) {
...
+                       kTestEvergreenVersion2 + strlen(kTestEvergreenVersion2));
+  std::vector<char> v3(kMaxEgVersionLength);
+  ASSERT_EQ(0, CompareEvergreenVersion(&v1, &v3));
+  ASSERT_EQ(0, CompareEvergreenVersion(&v1, &v1));
+  ASSERT_EQ(-1, CompareEvergreenVersion(&v1, &v2));
+  v3.assign(kTestEvergreenVersion3,
+            kTestEvergreenVersion3 + strlen(kTestEvergreenVersion3));
+  ASSERT_EQ(1, CompareEvergreenVersion(&v3, &v2))
```
Comment (hlwarriner) [inline starboard/loader_app/slot_management_test.cc:506]: We should try to write unit tests that exercise the public APIs  instead of the private helpers: go/unit-testing-practices#public-apis.

I think we can try to do this later in a follow-up PR, though.

File: starboard/loader_app/slot_management.cc [hunk 5]
```diff
@@ -148,9 +242,56 @@ void* LoadSlotManagedLibrary(const std::string& app_key,
...
+      if (!ReadEvergreenVersion(&system_image_manifest_file_path,
+                                system_image_version.data(),
+                                kMaxEgVersionLength)) {
+        SB_LOG(WARNING)
+            << "Failed to read the Evergreen version of the system image";
+      }
+    }
+
+    if (CompareEvergreenVersion(&system_image_version, &current_version) > 0) {
+      if (ImRollForward(0) != IM_ERROR) {
```
Comment (hlwarriner) [inline starboard/loader_app/slot_management.cc:285]: It would be great if we could add a YTS test for this behavior. It seems difficult to do though since we don't have file system access from YTS (for example to update the system image to a new version, following an EG-powered update in an installation slot). Do you have any ideas for this?

**Review by hlwarriner — COMMENTED** [review]
The changes look good - just left a few style comments, mostly about the tests. Thanks for the solid unit test coverage.

File: starboard/loader_app/slot_management_test.cc (hunk 4 above)
Comment (yuying-y) [reply starboard/loader_app/slot_management_test.cc:506]: A follow-up PR SGTM.

File: starboard/loader_app/slot_management_test.cc (hunk 3 above)
Comment (yuying-y) [reply starboard/loader_app/slot_management_test.cc:498]: I'll consider it in the follow-up PR for unit tests.

**Review by hlwarriner — APPROVED** [review]

**Review by kaidokert — APPROVED** [review]

File: starboard/loader_app/slot_management_test.cc [hunk 6]
```diff
@@ -468,6 +489,81 @@ TEST_P(SlotManagementTest, BadSabi) {
...
+TEST_P(SlotManagementTest, CompareEvergreenVersion) {
+  if (!storage_path_implemented_) {
+    return;
+  }
+  std::vector<char> v1(kTestEvergreenVersion1,
+                       kTestEvergreenVersion1 + strlen(kTestEvergreenVersion1));
+  std::vector<char> v2(kTestEvergreenVersion2,
+                       kTestEvergreenVersion2 + strlen(kTestEvergreenVersion2));
+  std::vector<char> v3(kMaxEgVersionLength);
+  ASSERT_EQ(0, CompareEvergreenVersion(&v1, &v3));
```
Comment (y4vor) [inline starboard/loader_app/slot_management_test.cc:501]: Please add a test with empty version. I think your code just logs a warning when it fails to read the version and keeps going and tries to use it for comparision.

**Review by y4vor — APPROVED** [review]

File: starboard/loader_app/slot_management_test.cc (hunk 6 above)
Comment (yuying-y) [reply starboard/loader_app/slot_management_test.cc:501]: v3 is an empty version, and it's used in the test case. CompareEvergreenVersion() handles empty versions, and returns 0, which doesn't interfere with the following code in LoadSlotManagedLibrary().
